### DISCUSSION TITLE
How to change the joint limits?

### DISCUSSION THREAD

**Discussion by mariogpascual** [upkie/upkie#122] [Software] [answered]
Hi, I wanted to ask how to change the current joint limits of Upkie and what is the rationale behind the current values. I think I only access these values through `box_[position, velocity, torque]_limits` functions, so would changing the URDF in `upkie_description` be enough to perform this modification?

Thanks!

**Comment by stephane-caron** [comment]
Changing the URDF is the expected way update the limits. We need to take care to modify the limits both on the C++ and Python side:

- The C++ side, used by the Bullet simulation spine, loads the URDF from the ``@upkie_description`` dependency. This dependency is defined in [`tools/workspace/upkie_description/repository.bzl`](https://github.com/tasts-robots/upkie/blob/f00a95a3bf2f036e571a26e609be5e2c18ad8226/tools/workspace/upkie_description/repository.bzl). To update it, fork the `upkie_description` repository, commit your modifications, then update the `remote` and `commit` fields of this file accordingly.
- The Python side, used by the `UpkieServosEnv` environment, actually loads the URDF description from PyPI (at [this line](https://github.com/tasts-robots/upkie/blob/f00a95a3bf2f036e571a26e609be5e2c18ad8226/upkie/envs/BUILD#L42) of the BUILD file). This is not convenient to carry around some local updates.

The current values were selected based on two points:

- For knee position limits: wheels touching the hips.
- For hip position limits: no good reason. I will update them to $[-0.5, 0.5]$ revolutions :wink: (Cables are the main limiters for hip positions, as they get fully stretched at $0.5$ rev.)

@mariogpascual Would that hip-limit update be sufficient to address your use case?

**Reply by stephane-caron** [reply]
Opening an issue about this at https://github.com/tasts-robots/upkie_description/issues/7

**Reply by stephane-caron** [reply, marked as answer]
Joint limits are now detailed in the `upkie_description` wiki: [Joint limits](https://github.com/upkie/upkie_description/wiki/Joint-limits).
